### PR TITLE
ZIP-244 & ZIP-245: creation of non-malleable txids + TZE signature hashing

### PR DESCRIPTION
**Pull request by nuttycom**

### CODE REVIEW

File: zcash_primitives/src/transaction/sighash_v5.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                precondition,
+                value,
+            } if self.txversion == TxVersion::ZFuture => {
+                let tzein = &tze_inputs[index];
+
+                tzein.prevout.write(&mut ch).unwrap();
+                CompactSize::write(&mut ch, precondition.extension_id.try_into().unwrap()).unwrap();
+                CompactSize::write(&mut ch, precondition.mode.try_into().unwrap()).unwrap();
+                Vector::write(&mut ch, &precondition.payload, |w, e| w.write_u8(*e)).unwrap();
+                ch.write_all(&value.to_i64_le_bytes()).unwrap();
```
Comment (nuttycom) [inline zcash_primitives/src/transaction/sighash_v5.rs:128]: @str4d now that I'm revisiting this signature hash code that we originally worked on together, I'm puzzled here - why is there a value associated with the TZE input?

Comment (str4d) [reply zcash_primitives/src/transaction/sighash_v5.rs:128]: The logic for adding it was the same as for transparent inputs: without it, there's nothing in the spend that directly commits to the value of the spent output, so a hardware wallet can't verify the fee without being provided those input coins. By having the value be part of the signature digest, the hardware wallet can present accurate information without needing the coin being spent, and the resulting signature can only be used if the client wasn't lying about the coin's value.

This is also the thing I was saying we should check against the hardware wallet fee vulnerability from last year, which IIRC said that this defense was ineffective if the adversary could ask the user to sign multiple times.

Comment (nuttycom) [reply zcash_primitives/src/transaction/sighash_v5.rs:128]: Ah, okay. I understand this better now than the last time we had this dicussion.

**Comment by nuttycom** [comment]
A question about this ZIP arose this morning in a discussion between myself and @str4d. The current draft of this ZIP specifies that the consensus_branch_id be committed to as part of the personalization string for transaction identifier hashes. This could potentially pose a problem for off-chain protocols which require transaction identifiers to remain stable across network upgrade boundaries, such as BOLT.  

One possible solution to this is to add the consensus_branch_id as an optional field to the transaction and include it as effecting data, such that a TxId may optionally (and would by default) commit to a consensus branch ID after which it would be considered invalid. This would make it possible to have transaction identifiers that remain stable across network upgrade boundaries, at the cost of replay protection. What do folks think here? Do protocols that depend on the consensus branch ID have to shut down or automatically terminate at network upgrade boundaries, or is there a better way?

cc/ @daira @dconnolly

**Comment by nuttycom** [comment]
Oops, that comment was supposed to go on the ZIP.

**Comment by daira** [comment]
It was a deliberate design decision to prioritize replay protection ahead of allowing transactions to remain valid across a network upgrade, when we designed Overwinter. I'm not saying we have to maintain that decision for all time; just pointing out that it wasn't something that happened accidentally.

**Comment by nuttycom** [comment]
subsumed by #375

**Review by Baiju311 — COMMENTED** [review]
Deleted due to spam comment.

**Pull request closed without merging**